src/tests/DemuxerTests.cpp

```diff
@@ -0,0 +1,174 @@
+public:
+  void parse_config(std::string const &ConfigurationStream,
+                    std::string const &ConfigurationModule) override {}
+  InitResult init_hdf(hdf5::node::Group &HDFGroup,
+                      std::string const &HDFAttributes) override {
+    return InitResult::OK();
+  }
+  InitResult reopen(hdf5::node::Group &HDFGrup) override {
+    return InitResult::OK();
+  }
+  WriteResult write(FlatbufferMessage const &Message) override {
+    return WriteResult::OK();
+  }
+  std::int32_t flush() override { return 0; }
+
+  std::int32_t close() override { return 0; }
+
+  void enable_cq(CollectiveQueue *cq, HDFIDStore *hdf_store,
+                 int mpi_rank) override {}
+};
+
+TEST_F(MessageTimeExtractionTest, Success) {
+  std::string TestKey("temp");
+  {
+    FlatbufferReaderRegistry::Registrar<DemuxerDummyReader2> RegisterIt(
+        TestKey);
+  }
+  char *TestData = new char[8];
+  std::memcpy(TestData + 4, TestKey.c_str(), 4);
+  FileWriter::FlatbufferMessage CurrentMessage(TestData, 8);
+  EXPECT_TRUE(CurrentMessage.isValid());
+  EXPECT_EQ(CurrentMessage.getSourceName(), "SomeSourceName");
+  EXPECT_EQ(CurrentMessage.getTimestamp(), 42ul);
+}
+
+TEST_F(MessageTimeExtractionTest, WrongFlatbufferID) {
+  std::string TestKey("temp");
+  std::string MessageKey("temr");
+  {
+    FlatbufferReaderRegistry::Registrar<DemuxerDummyReader2> RegisterIt(
+        TestKey);
+  }
+  char *TestData = new char[8];
+  std::memcpy(TestData + 4, MessageKey.c_str(), 4);
+
+  EXPECT_THROW(FileWriter::FlatbufferMessage(TestData, 8), UnknownFlatbufferID);
+}
+
+TEST_F(MessageTimeExtractionTest, WrongMsgSize) {
+  std::string TestKey("temp");
+  {
+    FlatbufferReaderRegistry::Registrar<DemuxerDummyReader2> RegisterIt(
+        TestKey);
+  }
+  char *TestData = new char[8];
+  std::memcpy(TestData + 4, TestKey.c_str(), 4);
+  EXPECT_THROW(FileWriter::FlatbufferMessage(TestData, 7), BufferTooSmallError);
+}
+
+class DemuxerTest : public ::testing::Test {
+public:
+  void SetUp() override {
+    std::map<std::string, ReaderPtr> &Readers =
+        FlatbufferReaderRegistry::getReaders();
+    Readers.clear();
+  }
+};
+
+TEST_F(DemuxerTest, Success) {
+  std::string TestKey("temp");
+  std::string SourceName("SomeSourceName");
+  {
+    FlatbufferReaderRegistry::Registrar<DemuxerDummyReader2> RegisterIt(
+        TestKey);
+  }
+  char *TestData = new char[8];
+  std::memcpy(TestData + 4, TestKey.c_str(), 4);
+  FileWriter::FlatbufferMessage CurrentMessage(TestData, 8);
+  ProcessMessageResult Result;
+  DemuxTopic TestDemuxer("SomeTopicName");
+  DummyWriter::ptr Writer(new DummyWriter);
+  Source DummySource(SourceName, TestKey, std::move(Writer));
+  TestDemuxer.add_source(std::move(DummySource));
+  ASSERT_EQ(TestDemuxer.sources().size(), size_t(1));
+  EXPECT_NO_THROW(TestDemuxer.sources().at(SourceName));
+  EXPECT_NO_THROW(Result =
+                      TestDemuxer.process_message(std::move(CurrentMessage)));
```

Comment:
Shouldn't use `std::move` here, the parameter is used as `const &` so it will have no effect.

Reply:
(repeated in a few tests)

Reply:
Done.